src/transform/src/reduction_pushdown.rs

```diff
@@ -81,8 +90,707 @@ impl ReductionPushdown {
+    /// Contains ((position of aggregation, the aggregation), filters)
+    single_input_aggs_to_push: Vec<(Vec<(usize, AggregateExpr)>, Vec<MirScalarExpr>)>,
+    /// Aggregations involving no inputs. They can be partially pushed down to
+    /// any input as long as the join key is not a primary key
+    /// Contains (position of aggregation, the aggregation)
+    no_input_aggs_to_push: Vec<(usize, AggregateExpr)>,
+
+    // Fields used to determine whether aggregations should not be pushed down
+    // into a particular input
+    /// Whether the join implementation is DeltaQuery. Avoid pushdowns that
+    /// would invalidate the DeltaQuery plan.
+    is_delta: bool,
+    /// The unique keys of each of the old inputs. Avoid pushdowns where there
+    /// exists a unique key that is a subset of the group keys
+    old_input_keys: Vec<Vec<Vec<usize>>>,
+
+    // Other helper fields
+    /// The equivalences of the old join, with all single-input filters removed
+    equivalences: Vec<Vec<MirScalarExpr>>,
+    /// The support of aggregations that we do not try to push down, sorted by
+    /// input. Inner reduces must not cause these columns to be removed from
+    /// the join.
+    aggs_not_pushed_support: Vec<HashSet<usize>>,
+    /// The mapping of global to local columns in the old join
+    old_join_mapper: JoinInputMapper,
+    /// For each input, stores the inputs it is bound to and the (global) join
+    /// keys that bind it to the particular input. This is a `BTreeMap` so that
+    /// the resulting plan is deterministic.
+    join_graph: Vec<BTreeMap<usize, Vec<MirScalarExpr>>>,
+    /// whether aggregations were pushed down to any input
+    pushdown_succeeded: bool,
+}
+
+impl<'a> ReductionPusher<'a> {
+    /// Creates a struct for pushing down reductions. Returns `None` if the
+    /// relation is ineligible for reduction pushdown.
+    fn try_new(
+        group_key: &'a [MirScalarExpr],
+        aggregates: &[AggregateExpr],
+        inputs: &[MirRelationExpr],
+        equivalences: &'a [Vec<MirScalarExpr>],
+        implementation: &expr::JoinImplementation,
+        predicates: &[MirScalarExpr],
+    ) -> Option<Self> {
+        let input_types = inputs.iter().map(|i| i.typ()).collect::<Vec<_>>();
+        let old_join_mapper = JoinInputMapper::new_from_input_types(&input_types);
+
+        // sort aggregations by how and whether they can be pushed down
+        // TODO: attempt to rewrite multi-input aggregations in terms of a
+        // single input
+        // TODO: attempt to reduce the number of inputs that aggregations come
+        // from by rewriting aggregations on join keys in terms of the
+        // equivalent expression from the other input
+        let mut outer_aggs = Vec::new();
+        let mut single_input_aggs_to_push = vec![(Vec::new(), Vec::new()); inputs.len()];
+        let mut no_input_aggs_to_push = Vec::new();
+
+        // classify the aggregations by how many inputs each aggregation references
```

Comment:
I think the comment here may be stale. This seems to be about classifying aggregations based on whether they can be pushed down or not (for various reasons that include the number of inputs referenced).